library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java

```diff
@@ -51,291 +66,607 @@
+          wal.windowWalParts.put(lastWindow, wal.walEndPointerAfterRecovery.getPartNum());
+          return lastWindow;
+        }
+      }
     }
-    return data;
+    return Stateless.WINDOW_ID;
   }
 
-  @Override
-  public long[] getWindowIds(int operatorId) throws IOException
+  /**
+   * Helper method that catches IOException while reading from wal to check if an entry was saved completely or not.
+   * @param reader wal reader
+   * @return wal entry
+   */
+  private Slice readNext(FileSystemWAL.FileSystemWALReader reader)
   {
-    Path operatorPath = new Path(appPath, String.valueOf(operatorId));
-    if (!fs.exists(operatorPath) || fs.listStatus(operatorPath).length == 0) {
+    try {
+      return reader.next();
+    } catch (IOException ex) {
+      //exception while reading wal entry which can be because there may have been failure while persisting an
+      //artifact so this window is not a finished window.
+      try {
+        reader.close();
+      } catch (IOException ioe) {
+        //closing the reader quietly.
+      }
       return null;
     }
-    return storageAgent.getWindowIds(operatorId);
   }
 
-  @Override
-  public long[] getWindowIds() throws IOException
+  /**
+   * Helper method that catches IOException while skipping an entry from wal to check if an entry was saved
+   * completely or not.
+   * @param reader wal reader
+   * @return true if skip was successful; false otherwise.
+   */
+  private boolean skipNext(FileSystemWAL.FileSystemWALReader reader)
   {
-    SortedSet<Long> windowIds = replayState.keySet();
-    long[] windowIdsArray = new long[windowIds.size()];
-
-    int index = 0;
+    try {
+      reader.skipNext();
+      return true;
+    } catch (IOException ex) {
+      //exception while skipping wal entry which can be because there may have been failure while persisting an
+      //artifact so this window is not a finished window.
+      try {
+        reader.close();
+      } catch (IOException e) {
+        //closing the reader quietly
+      }
+      return false;
+    }
+  }
 
-    for (Long windowId: windowIds) {
-      windowIdsArray[index] = windowId;
-      index++;
+  private void closeReaders() throws IOException
+  {
+    //close all reader stream and remove read-only wals
+    wal.getReader().close();
+    if (readOnlyWals.size() > 0) {
+      Iterator<Map.Entry<Integer, FSWindowReplayWAL>> walIterator = readOnlyWals.entrySet().iterator();
+      while (walIterator.hasNext()) {
+        Map.Entry<Integer, FSWindowReplayWAL> entry = walIterator.next();
+        entry.getValue().getReader().close();
+
+        int operatorId = entry.getKey();
+        if (deletedOperators == null || !deletedOperators.contains(operatorId)) {
+          //the read only wal can be removed.
+          walIterator.remove();
+        }
+      }
     }
+  }
 
-    return windowIdsArray;
+  @Override
+  public void save(Object object, long windowId) throws IOException
```

Comment:
JMS Input in its endWindow does the following:
1. saves the artifact
2. acks the session

When acks fails but the artifact got saved, then it tries to delete the artifact.
If the artifact is deleted successfully, the window is considered completed or not.

1. When deleted - window was not completed and the messages will be consumed from JMS
2. When not deleted - window was considered completed so it is replayed from file. However while replaying message ids are added to pendingAck set. When these messages are received from JMS again (since ack failed earlier), they will be considered re-delivered.

Deletion of a window during ```setup``` is difficult because we do not actually know if ack failed for that window.